core/trino-main/src/main/java/io/trino/operator/exchange/ScaleWriterPartitioningManager.java

```diff
@@ -0,0 +1,362 @@
+        for (int i = 0; i < writerStats.length; i++) {
+            writerIdWithRowCount.add(new IdWithRowCount(i, writerStats[i].getRowCountSinceLastScaleUp()));
+        }
+
+        Set<Integer> scaledPartitionIds = new HashSet<>();
+
+        // Find skewed partitions and scale them across multiple writers
+        while (!writerIdWithRowCount.isEmpty()) {
+            // Find the writer with maximum row count
+            IdWithRowCount maxWriter = writerIdWithRowCount.last();
+            writerIdWithRowCount.remove(maxWriter);
+
+            // Find the list of writers with are skewed against the writer with maximum row count
+            List<IdWithRowCount> minWriters = findSkewedMinWriters(maxWriter, writerIdWithRowCount);
+
+            if (minWriters.isEmpty()) {
+                break;
+            }
+
+            WriterStats maxWriterStats = writerStats[maxWriter.id];
+            // Find the partition with maximum row count in the skewed writer, while skipping the partitions which are
+            // already scaled in this cycle.
+            Optional<IdWithRowCount> partitionWithMaxWriterCount = maxWriterStats.getPartitionWithMaxWriterCount(scaledPartitionIds);
+            if (partitionWithMaxWriterCount.isEmpty()) {
+                continue;
+            }
+
+            boolean isScaled = false;
+            IdWithRowCount partitionToScale = partitionWithMaxWriterCount.get();
+            // Using skewed writer physicalWrittenBytes and total row count, estimate the partition physical written
+            // bytes since last scale up.
+            long estimatedPartitionPhysicalWrittenBytes = maxWriterStats.estimatePartitionPhysicalWrittenBytes(
+                    partitionToScale.rowCount,
+                    writerPhysicalWrittenBytes.get(maxWriter.id));
+            PartitionWriterInfo partitionWriterInfo = partitionWriterInfos[partitionToScale.id];
+
+            // Calculate delta row count which will be used to calculate partition row count after scaling
+            long deltaRowCount = partitionToScale.rowCount / 2;
+
+            for (IdWithRowCount minWriter : minWriters) {
+                // Try scaling the partition to this minimum skewed writer
+                if (!isScaled && scalePartition(partitionToScale.id, minWriter.id, partitionWriterInfo, estimatedPartitionPhysicalWrittenBytes)) {
+                    scaledPartitionIds.add(partitionToScale.id);
+                    // If scaling happened then add the minimum skewed writer to the sorted set while accounting for
+                    // scaled partition row count.
+                    writerIdWithRowCount.add(new IdWithRowCount(minWriter.id, minWriter.rowCount + deltaRowCount));
+                    isScaled = true;
+                }
+                else {
+                    // If scaling isn't possible with this minimum skewed writer then simply add the writer back to sorted set
```

Comment:
When this could happen (e.g. both max and min writer write same partition)? In such case shouldn't you just pick another partition from `maxNode` to rebalance?